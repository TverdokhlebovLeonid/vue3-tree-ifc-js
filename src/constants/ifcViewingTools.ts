export enum IFC_VIEWING_TOOLS {
  createPlane = 'CREATE_PLANE',
  createCoordinates = 'CREATE_COORDINATES',
  cancelPlane = 'CANCEL_PLANE',
  navCube = 'NAV_CUBE',
  fullScreen = 'FULL_SCREEN',
}
