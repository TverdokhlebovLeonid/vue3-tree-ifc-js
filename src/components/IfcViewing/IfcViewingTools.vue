<script setup lang="ts">
import { ref, reactive, computed, onUnmounted } from 'vue'
import { TOOLS } from '@/components/IfcViewing/dataIfcViewing'
import type { ITools } from '@/components/IfcViewing/interfaceIfcViewing'
import { IFC_VIEWING_TOOLS } from '@/constants/ifcViewingTools'

const emits = defineEmits<{
  (e: 'open-fullscreen'): void
  (e: 'tool-selection', value: string): void
}>()

const toolsButton = reactive<ITools[]>(TOOLS)
const activeTool = ref<string>('')
const isCancelAction = computed((): boolean => activeTool.value === IFC_VIEWING_TOOLS.createPlane)

const setActiveTool = (tool: string = ''): void => {
  toolsButton.forEach((el) => {
    el.active = el.tool === tool
  })
}
const startStateTools = (): void => {
  activeTool.value = ''
  setActiveTool()
  emits('tool-selection', '')
}
const setTools = (tool: string): void => {
  if (tool === activeTool.value) {
    startStateTools()
    return
  }
  if (tool === IFC_VIEWING_TOOLS.fullScreen) {
    setFullScreen()
    return
  }
  activeTool.value = tool
  setActiveTool(tool)
  emits('tool-selection', tool)
}

const cancelTool = (): void => {
  emits('tool-selection', IFC_VIEWING_TOOLS.cancelPlane)
}

const setFullScreen = (): void => {
  emits('open-fullscreen')
}

defineExpose({ startStateTools, setTools, activeTool, cancelTool, setFullScreen })
onUnmounted(() => {
  startStateTools()
})
</script>

<template>
  <div class="ifc-tools">
    <div class="ifc-tools__cancel">
      <el-tooltip
        v-if="isCancelAction"
        effect="dark"
        content="Отменить действие инструмента"
        placement="bottom"
      >
        <el-button type="danger" size="small" plain icon="Failed" @click="cancelTool">
          Отмена
        </el-button>
      </el-tooltip>
    </div>
    <div class="ifc-tools__buttons">
      <template v-for="tool in toolsButton" :key="tool.icon">
        <el-tooltip
          effect="dark"
          :content="tool.active ? 'Отключить инструмент' : tool.popover"
          placement="bottom"
        >
          <el-button
            type="primary"
            :class="{ 'ifc-tools__buttons_active': tool.active }"
            plain
            size="small"
            :icon="tool.icon"
            @click="setTools(tool.tool)"
          >
            {{ tool.name }}
          </el-button>
        </el-tooltip>
      </template>
    </div>
  </div>
</template>

<style scoped lang="sass">
.ifc-tools
  margin-top: 80px
  display: flex
  flex-direction: row
  justify-content: space-between
  &__buttons
    display: flex
    flex-direction: row
    justify-content: flex-end
    &_active
      background-color: $color-btn-active
</style>
