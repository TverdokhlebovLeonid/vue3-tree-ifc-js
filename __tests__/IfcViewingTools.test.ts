import { mount } from '@vue/test-utils'
import { expect } from 'vitest'
import IfcViewingTools from '@/components/IfcViewing/IfcViewingTools.vue'
import { IFC_VIEWING_TOOLS } from '@/constants/ifcViewingTools'
import type { ComponentWrapperType } from './interfaceIfcTests'

describe('IfcViewingTools test component', () => {
  let component: ComponentWrapperType<InstanceType<typeof IfcViewingTools>>

  beforeEach(() => {
    component = mount(IfcViewingTools)
  })

  it('Should render', async () => {
    expect(
      component
        .find('.ifc-tools')
        .find('.ifc-tools__buttons')
        .find('.el-button')
        .find('.el-icon')
        .exists(),
    ).toBe(true)
    expect(
      component
        .find('.ifc-tools__buttons')
        .findAll('span')
        .map((span) => span.text()),
    ).toEqual(['Плоскость', 'Координаты', 'Куб', 'Во весь экран'])
    const button = component.find('button')
    await button.trigger('click')
    expect(component.emitted('tool-selection')).toBeTruthy()
  })

  it('Snapshot component', () => {
    expect(component.html()).toMatchSnapshot()
  })

  it('Function setTools', () => {
    component.vm.setTools(IFC_VIEWING_TOOLS.createPlane)
    expect(component.emitted('tool-selection')?.[0]).toEqual([IFC_VIEWING_TOOLS.createPlane])
    expect(component.vm.activeTool).toBe(IFC_VIEWING_TOOLS.createPlane)
  })

  it('Function setTools(fullScreen)', () => {
    const component = mount(IfcViewingTools)
    component.vm.setTools(IFC_VIEWING_TOOLS.fullScreen)
    expect(component.emitted('open-fullscreen')).toBeTruthy()
  })

  it('Function setTools(same tool)', () => {
    component.vm.activeTool = IFC_VIEWING_TOOLS.navCube
    component.vm.setTools(IFC_VIEWING_TOOLS.navCube)
    expect(component.emitted('tool-selection')?.[0]).toEqual([''])
    expect(component.vm.activeTool).toBe('')
  })

  it('Function startStateTools', () => {
    component.vm.startStateTools()
    expect(component.emitted('tool-selection')?.[0]).toEqual([''])
    expect(component.vm.activeTool).toBe('')
  })

  it('Function cancelTool', () => {
    component.vm.cancelTool()
    expect(component.emitted('tool-selection')?.[0]).toEqual([IFC_VIEWING_TOOLS.cancelPlane])
  })

  it('Function setFullScreen', async () => {
    component.vm.setFullScreen()
    expect(component.emitted('open-fullscreen')).toBeTruthy()
  })
})
