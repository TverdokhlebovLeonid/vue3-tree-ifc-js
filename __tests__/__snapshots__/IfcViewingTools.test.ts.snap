// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IfcViewingTools test component > Snapshot component 1`] = `
"<div data-v-b988461a="" class="ifc-tools">
  <div data-v-b988461a="" class="ifc-tools__cancel">
    <!--v-if-->
  </div>
  <div data-v-b988461a="" class="ifc-tools__buttons"><button data-v-b988461a="" aria-disabled="false" type="button" class="el-button el-button--primary el-button--small is-plain ifc-tools__buttons_active el-tooltip__trigger el-tooltip__trigger"><i class="el-icon">
        <coin></coin>
      </i><span class="">Плоскость</span></button>
    <!--teleport start-->
    <!--teleport end--><button data-v-b988461a="" aria-disabled="false" type="button" class="el-button el-button--primary el-button--small is-plain el-tooltip__trigger"><i class="el-icon">
        <coordinate></coordinate>
      </i><span class="">Координаты</span></button>
    <!--teleport start-->
    <!--teleport end--><button data-v-b988461a="" aria-disabled="false" type="button" class="el-button el-button--primary el-button--small is-plain el-tooltip__trigger"><i class="el-icon">
        <box></box>
      </i><span class="">Куб</span></button>
    <!--teleport start-->
    <!--teleport end--><button data-v-b988461a="" aria-disabled="false" type="button" class="el-button el-button--primary el-button--small is-plain el-tooltip__trigger"><i class="el-icon">
        <fullscreen></fullscreen>
      </i><span class="">Во весь экран</span></button>
    <!--teleport start-->
    <!--teleport end-->
  </div>
</div>"
`;
